### PR TITLE
Filter by dropdown selections

### LINKED ISSUE #125: Filter model list by the item selected in the filter UI
When clicking an item in the filter UI then filter the model list.

Blocked by https://github.com/canonical-web-and-design/jaas-dashboard/issues/124

### PR DESCRIPTION
**Pull request by hatched**
## Done

When selecting filter options from the drop down filter the table list.

## QA

- Pull code
- Run `./run clean && ./run serve`
- Open http://0.0.0.0:8036/
- View the model list and select some items from the filter dropdown, it should filter the resulting mode list.
- Switch between the groupings (status, cloud, owner) the filters should persist.
- Refresh, the filters should persist.

## Details

Fixes #125

### CODE REVIEW

File: src/app/selectors.js [hunk 1]
```diff
@@ -421,26 +513,6 @@ export const getGroupedApplicationsDataByStatus = createSelector(
   groupApplicationsByStatus
 );
 
-/**
```
Comment (hatched) [inline src/app/selectors.js]: These are no longer used anywhere and were removed.

File: src/components/FilterTags/FilterTags.js [hunk 2]
```diff
@@ -20,9 +20,14 @@ const FilterTags = () => {
   const history = useHistory();
   const location = useLocation();
 
-  const queryParams = queryString.parse(location.search);
-  const queryParamsActiveFilters = queryParams.activeFilters;
-
+  const queryParams = queryString.parse(location.search, {
```
Comment (hatched) [inline src/components/FilterTags/FilterTags.js]: These changes were a drive-by to remove the query string key duplication.

File: src/components/FilterTags/FilterTags.js [hunk 3]
```diff
@@ -179,7 +186,7 @@ const FilterTags = () => {
             activeFilters.length > 0 &&
             activeFilters.map(activeFilter => (
               <span className="p-filter-tags__active-filter" key={activeFilter}>
-                {activeFilter}
+                {activeFilter.replace(":", ": ")}
```
Comment (hatched) [inline src/components/FilterTags/FilterTags.js]: This was another drive-by to remove the space in the filter values to make for easier parsing. The UI maintains the space.

**Comment by barrymcgee** [comment]
Feels like we should display a message if the filters exclude all models?

Screenshot 2020-02-19 at 20 01 37

**Comment by hatched** [comment]
You mean like this? https://github.com/canonical-web-and-design/jaas-dashboard/issues/122

File: src/app/selectors.js [hunk 4]
```diff
@@ -307,6 +320,61 @@ const countModelStatusGroups = groupedModelStatuses => {
...
+              data.info.cloudCredentialTag.split("@")[1].split("_")[1]
+            );
+          }
+          break;
+        case "region":
+          return !values.includes(data.model.region);
+        case "owner":
+          if (data.info) {
+            return !values.includes(
+              data.info.ownerTag.split("@")[0].replace("user-", "")
```
[the remark this answers is not in the document]
Comment (hatched) [reply src/app/selectors.js]: done

File: src/app/selectors.js [hunk 5]
```diff
@@ -307,6 +320,61 @@ const countModelStatusGroups = groupedModelStatuses => {
...
+    filterSegments[values[0]].push(values[1]);
+  });
+
+  const clonedModelData = cloneDeep(modelData);
+
+  Object.entries(clonedModelData).forEach(([uuid, data]) => {
+    const remove = Object.entries(filterSegments).some(([segment, values]) => {
+      switch (segment) {
+        case "cloud":
+          return !values.includes(data.model.cloudTag.replace("cloud-", ""));
```
[the remark this answers is not in the document]
Comment (hatched) [reply src/app/selectors.js]: done

File: src/app/selectors.js [hunk 6]
```diff
@@ -307,6 +320,61 @@ const countModelStatusGroups = groupedModelStatuses => {
...
+        case "credential":
+          if (data.info) {
+            return !values.includes(
+              data.info.cloudCredentialTag.split("@")[1].split("_")[1]
```
[the remark this answers is not in the document]
Comment (hatched) [reply src/app/selectors.js]: done

**Comment by spencerbygraves** [comment]
When I select a filter, my instinct is to click it again to deselect it. Especially when you select something at the bottom of the list, then have to move to the top to remove it.

@ziheliu214 also made a good suggestion; do we need a way to remove all the filters at once?

Something to think about @clagom?

**Comment by hatched** [comment]
@spencerbygraves feel free to create a discussion issue in the squad repo.
